contents/python/numpy.md

```diff
@@ -0,0 +1,167 @@
+```python
+>>> np.array([1.5, "a"])
+array(['1.5', 'a'], dtype='<U32') # not a numeric type!
+>>> np.array([10 ** 20, 1])
+array([100000000000000000000, 1], dtype=object)
+```
+
+## The array type
+
+NumPy is built around its `array` type (an alias for `ndarray`, hinting at its multidimensional nature). Unlike Python
+lists, elements of `array` instances are stored in memory contiguously, reducing interpreter overhead
+and enabling parallel operations.
+
+An `array` object can be initialised from iterables or nested lists through `np.array` (assuming NumPy has been
+imported as `np`, as is usually done in production code, and which will be assumed in the rest of this article):
+```python
+a = np.array([7, 3, 9, 3, 1, 3, 3, 7]) # 1D array
+b = np.array([[8, 1, 6], [3, 5, 7], [4, 9, 2]]) # 2D array
+```
+By default, these arrays are indexed in C order (last index changes fastest), as can be illustrated by
+invoking the `reshape` array method:
+```python
+>>> np.array(range(27)).reshape((3,3,3))
+array([[[ 0,  1,  2],
+        [ 3,  4,  5],
+        [ 6,  7,  8]],
+
+       [[ 9, 10, 11],
+        [12, 13, 14],
+        [15, 16, 17]],
+
+       [[18, 19, 20],
+        [21, 22, 23],
+        [24, 25, 26]]])
+```
+Otherwise, indexing is mostly as in standard Python, with the convenience that extra square brackets can be
+omitted for multi-_axis_ (array coordinate) indexing, so `a[0][1:][2]` becomes `a[0,1:,2]`.
+See [this page](https://numpy.org/devdocs/reference/arrays.indexing.html) for a more detailed explanation.
+
+## Library functions
+
+In addition to its array type, NumPy provides a large [library](https://numpy.org/devdocs/reference/routines.html)
+for manipulating arrays. These include:
+* [creating special types of arrays](https://numpy.org/devdocs/reference/routines.array-creation.html)
+* [sorting and searching](https://numpy.org/devdocs/reference/routines.sort.html)
+* [random number generation](https://numpy.org/devdocs/reference/random/index.html)
+* [string operations](https://numpy.org/devdocs/reference/routines.char.html), with strings treated as character arrays
+* [linear algebra](https://numpy.org/devdocs/reference/routines.linalg.html) and other
+[mathematical functions](https://numpy.org/devdocs/reference/routines.math.html)
+
+```python
+# Example: solve a system of linear equations with NumPy
+>>> A = np.array([[9, -9, -10], [8, 5, -4], [0, 7, -9]])
+>>> b = np.array([-3, 5, -7])
+>>> np.linalg.solve(A, b)
+array([0.95958854, 0.22997796, 0.95664952])
+```
+
+Many of these functions work equally well on one-dimensional arrays as they do on 11-dimensional ones, and their
+scope can be refined by passing axis arguments as in the [array average example](#what-numpy-can-and-cannot-do-for-you)
+above. In some cases, the function is accelerated further through
+[universal functions](https://numpy.org/devdocs/reference/ufuncs.html), vectorised C code with thin Python wrappers.
```

Comment:
Given that the reader may only have basic Python knowledge and new to scientific computing, try to reduce jargon or scientific terms if they are not strictly necessary.